crates/core/src/factories/core_space/test.rs

```diff
@@ -1,56 +1,89 @@
-use kitsune2_api::*;
+use kitsune2_api::{kitsune::*, space::*, *};
+use kitsune2_test_utils::agent::*;
 use std::sync::{Arc, Mutex};
 
-const SIG: &[u8] = b"fake-signature";
+#[tokio::test(flavor = "multi_thread")]
+async fn space_local_agent_join_leave() {
+    #[derive(Debug)]
+    struct S;
 
-#[derive(Debug)]
-struct TestCrypto;
+    impl SpaceHandler for S {}
 
-impl agent::Signer for TestCrypto {
-    fn sign(
-        &self,
-        _agent_info: &agent::AgentInfo,
-        _encoded: &[u8],
-    ) -> BoxFut<'_, K2Result<bytes::Bytes>> {
-        Box::pin(async move { Ok(bytes::Bytes::from_static(SIG)) })
-    }
-}
+    #[derive(Debug)]
+    struct K;
 
-impl agent::Verifier for TestCrypto {
-    fn verify(
-        &self,
-        _agent_info: &agent::AgentInfo,
-        _message: &[u8],
-        signature: &[u8],
-    ) -> bool {
-        signature == SIG
+    impl KitsuneHandler for K {
+        fn create_space(
+            &self,
+            _space: SpaceId,
+        ) -> BoxFut<'_, K2Result<space::DynSpaceHandler>> {
+            Box::pin(async move {
+                let s: DynSpaceHandler = Arc::new(S);
+                Ok(s)
+            })
+        }
     }
-}
 
-const S1: SpaceId = SpaceId(id::Id(bytes::Bytes::from_static(b"space1")));
-
-fn make_agent_info(id: AgentId, url: Url) -> Arc<agent::AgentInfoSigned> {
-    let created_at = Timestamp::now();
-    let expires_at = created_at + std::time::Duration::from_secs(60 * 20);
-    futures::executor::block_on(agent::AgentInfoSigned::sign(
-        &TestCrypto,
-        agent::AgentInfo {
-            agent: id,
-            space: S1.clone(),
-            created_at,
-            expires_at,
-            is_tombstone: false,
-            url: Some(url),
-            storage_arc: DhtArc::FULL,
-        },
-    ))
+    let k: DynKitsuneHandler = Arc::new(K);
+    let k1 = builder::Builder {
+        verifier: Arc::new(TestVerifier),
+        ..crate::default_builder()
+    }
+    .with_default_config()
     .unwrap()
+    .build(k)
+    .await
+    .unwrap();
+
+    let bob = Arc::new(TestLocalAgent::default()) as agent::DynLocalAgent;
+    let ned = Arc::new(TestLocalAgent::default()) as agent::DynLocalAgent;
+
+    let s1 = k1.space(TEST_SPACE.clone()).await.unwrap();
+
+    s1.local_agent_join(bob.clone()).await.unwrap();
+    s1.local_agent_join(ned.clone()).await.unwrap();
+
+    let mut active_peer_count = 0;
+
+    for _ in 0..5 {
+        active_peer_count = 0;
+        for peer in s1.peer_store().get_all().await.unwrap() {
+            if !peer.is_tombstone {
+                active_peer_count += 1;
+            }
+        }
+        if active_peer_count == 2 {
+            break;
+        }
+        tokio::time::sleep(std::time::Duration::from_millis(100)).await;
```

Comment:
These are accumulating a bit... I really like how fast the tests are at the moment and no matter how slow it gets to run them all, I'll be able to run per module so it won't be too bad.

Is it worth the effort to think about how else this could be tested without a sleep?

Reply:
Maybe a little bit better at least? https://github.com/holochain/kitsune2/pull/55/commits/f5120be2a6969b652225d5c17dabc066e9bc448f

Reply:
Yes that is better, I'm happy to trade a bit of extra heat for faster tests :)